File: ScriptsToProcess/NEW BATCH TO PROCESS/ColorMeshTiles.py
"""
COLORED MESH TILES: Branch-aware idempotent production engine.
Consolidates Mega-Meshes per surface, tracks discrete materials, 
and binds surfaces together into selectable Rhino Groups.
Runtime: Rhino 8 Python 3 (CPython)
"""

import Rhino
import Rhino.Geometry as rg
import System.Drawing as sd
import Grasshopper
import random
import time

# --- Telemetry Clock Start ---
t_start = time.perf_counter()

# --- Metadata Initialization ---
ghenv.Component.Name = "ColorMeshTiles"
ghenv.Component.NickName = "ColorMeshTiles"
ghenv.Component.Description = "BIM-ready Idempotent Engine. Consolidates Mega-Meshes per surface and binds them into Rhino Groups."

# Pre-allocate output tree structures safely
panel_mesh = Grasshopper.DataTree[object]()
panel_colors = Grasshopper.DataTree[object]()
panel_tags = Grasshopper.DataTree[object]()
panel_geometry = Grasshopper.DataTree[object]()

# --- Graceful Validation Matrix ---
inputs_valid = True
if 'polylines' not in globals() or polylines is None or polylines.BranchCount == 0:
    inputs_valid = False
if 'gradient_colors' not in globals() or not gradient_colors:
    inputs_valid = False

if not inputs_valid:
    ghenv.Component.Message = "COLORED_TILES\nSTATUS: IDLE\n---\nAWAITING DATA"
else:
    random.seed(42)
    
    axis_idx = axis if 'axis' in globals() and axis is not None else 2
    j_factor = max(0.0, min(100.0, jitter_pct)) / 100.0 if 'jitter_pct' in globals() and jitter_pct is not None else 0.0
    a_factor = max(0.0, min(100.0, accent_pct)) / 100.0 if 'accent_pct' in globals() and accent_pct is not None else 0.0
    i_factor = max(0.0, min(1.0, inset_factor)) if 'inset_factor' in globals() and inset_factor is not None else 1.0
    
    do_bake = bool(bake_trigger) if 'bake_trigger' in globals() and bake_trigger is not None else False
    b_name = str(bake_name) if 'bake_name' in globals() and bake_name is not None else ""
    
    num_available_colors = len(gradient_colors)
    color_to_tag = {gradient_colors[i]: f"Tile {i + 1}" for i in range(num_available_colors)}
    
    # Global tracking for UI Telemetry
    global_color_counts = {c: 0 for c in gradient_colors}
    has_accent = 'accent_color' in globals() and accent_color is not None and a_factor > 0
    if has_accent:
        global_color_counts[accent_color] = 0
        color_to_tag[accent_color] = "Accent Tile"

    # Allocation Trees
    out_mesh_tree = Grasshopper.DataTree[object]()
    out_cols_tree = Grasshopper.DataTree[object]()
    out_tags_tree = Grasshopper.DataTree[object]()
    out_geo_tree = Grasshopper.DataTree[object]()

    # ==========================================================================
    # PASS 1: GLOBAL BOUNDS SCAN
    # ==========================================================================
    global_keys = []
    for i in range(polylines.BranchCount):
        branch = polylines.Branch(i)
        for crv in branch:
            if crv is None: continue 
                
            bbox = crv.GetBoundingBox(True)
            global_keys.append(bbox.Center.X if axis_idx == 0 else (bbox.Center.Y if axis_idx == 1 else bbox.Center.Z))

    if not global_keys:
        ghenv.Component.Message = "HME\nSTATUS: EMPTY\n---\nNO VALID CURVES"
    else:
        global_min = min(global_keys)
        global_max = max(global_keys)
        coordinate_span = global_max - global_min if global_max != global_min else 1.0

        # ==========================================================================
        # PASS 2: BRANCH-AWARE CONSOLIDATION
        # ==========================================================================
        total_panels = 0

        for i in range(polylines.BranchCount):
            path = polylines.Path(i)
            branch = polylines.Branch(i)
            
            out_mesh_tree.EnsurePath(path)
            out_cols_tree.EnsurePath(path)
            out_tags_tree.EnsurePath(path)
            out_geo_tree.EnsurePath(path)
            
            branch_geometries = []
            
            # Initialize local mesh buckets strictly for this specific surface/branch
            local_mesh_buckets = {c: rg.Mesh() for c in gradient_colors}
            if has_accent:
                local_mesh_buckets[accent_color] = rg.Mesh()
            
            for crv in branch:
                if crv is None: continue
                
                success, polyline = crv.TryGetPolyline()
                if not success: continue
                
                bbox = crv.GetBoundingBox(True)
                center_pt = bbox.Center
                
                if i_factor < 1.0:
                    scale_transform = rg.Transform.Scale(center_pt, i_factor)
                    polyline.Transform(scale_transform)
                
                pts = list(polyline)
                if len(pts) < 4: continue
                if pts[0].EpsilonEquals(pts[-1], 1e-6):
                    pts.pop()
                
                num_vertices = len(pts)
                
                current_val = center_pt.X if axis_idx == 0 else (center_pt.Y if axis_idx == 1 else center_pt.Z)
                t_base = (current_val - global_min) / coordinate_span
                
                if j_factor > 0:
                    t_base += random.uniform(-j_factor * 0.5, j_factor * 0.5)
                
                t_base = max(0.0, min(0.999999, t_base))
                color_index = int(t_base * num_available_colors)
                cell_color = gradient_colors[color_index]
                
                if has_accent and random.random() < a_factor:
                    cell_color = accent_color
                
                global_color_counts[cell_color] += 1
                
                target_mesh = local_mesh_buckets[cell_color]
                v_start_idx = target_mesh.Vertices.Count
                
                target_mesh.Vertices.Add(center_pt)
                target_mesh.VertexColors.Add(cell_color)
                
                for p in pts:
                    target_mesh.Vertices.Add(p)
                    target_mesh.VertexColors.Add(cell_color)
                
                for j in range(num_vertices):
                    next_j = (j + 1) % num_vertices
                    target_mesh.Faces.AddFace(v_start_idx, v_start_idx + 1 + j, v_start_idx + 1 + next_j)
                
                branch_geometries.append(polyline.ToPolylineCurve())
                total_panels += 1
            
            # Compact and append local buckets to the Data Trees
            local_meshes = []
            local_colors = []
            local_tags = []
            
            for color, m_bucket in local_mesh_buckets.items():
                if m_bucket.Vertices.Count > 0:
                    m_bucket.Normals.ComputeNormals()
                    m_bucket.Compact()
                    local_meshes.append(m_bucket)
                    local_colors.append(color)
                    local_tags.append(color_to_tag[color])
                    
            if local_meshes:
                out_mesh_tree.AddRange(local_meshes, path)
                out_cols_tree.AddRange(local_colors, path)
                out_tags_tree.AddRange(local_tags, path)
            if branch_geometries:
                out_geo_tree.AddRange(branch_geometries, path)

        panel_mesh = out_mesh_tree
        panel_colors = out_cols_tree
        panel_tags = out_tags_tree
        panel_geometry = out_geo_tree

        # ==========================================================================
        # PASS 3: IDEMPOTENT BIM BAKING SUBROUTINE WITH GROUP BINDING
        # ==========================================================================
        bake_status = ""
        items_replaced = 0
        
        if do_bake:
            doc = Rhino.RhinoDoc.ActiveDoc
            
            # Search & Destroy Previous Iterations
            if b_name:
                existing_objs = doc.Objects.FindByUserString("ElefrontBakeName", b_name, False)
                if existing_objs:
                    for obj in existing_objs:
                        doc.Objects.Delete(obj.Id, True)
                        items_replaced += 1
            
            # Iterate through the synchronized Data Trees
            for i in range(out_mesh_tree.BranchCount):
                path = out_mesh_tree.Path(i)
                branch_meshes = out_mesh_tree.Branch(i)
                branch_colors = out_cols_tree.Branch(i)
                branch_tags = out_tags_tree.Branch(i)
                
                branch_id = str(path).strip("{}").replace(";", "_")
                
                # --- RHINO GROUP GENERATION ---
                # Safely search for an existing group or spawn a new one
                group_name = f"SurfaceGroup_{b_name}_{branch_id}" if b_name else f"SurfaceGroup_{branch_id}"
                group_idx = -1
                
                for g in doc.Groups:
                    if g and g.Name == group_name:
                        group_idx = g.Index
                        break
                        
                if group_idx < 0:
                    group_idx = doc.Groups.Add(group_name)
                
                # Assign Mega-Meshes to Layers and bind them into the surface Group
                for f_mesh, f_color, f_tag in zip(branch_meshes, branch_colors, branch_tags):
                    layer_name = f"HexFacade_{f_tag.replace(' ', '')}"
                    
                    layer_idx = doc.Layers.Find(layer_name, True)
                    if layer_idx < 0:
                        new_layer = Rhino.DocObjects.Layer()
                        new_layer.Name = layer_name
                        new_layer.Color = f_color
                        
                        new_mat = Rhino.DocObjects.Material()
                        new_mat.DiffuseColor = f_color
                        new_mat.Name = f"Mat_{layer_name}"
                        
                        mat_idx = doc.Materials.Add(new_mat)
                        new_layer.RenderMaterialIndex = mat_idx
                        layer_idx = doc.Layers.Add(new_layer)
                    
                    attr = Rhino.DocObjects.ObjectAttributes()
                    attr.LayerIndex = layer_idx
                    attr.ColorSource = Rhino.DocObjects.ObjectColorSource.ColorFromLayer
                    attr.MaterialSource = Rhino.DocObjects.ObjectMaterialSource.MaterialFromLayer
                    
                    # INJECT BIM METADATA
                    if b_name:
                        attr.SetUserString("ElefrontBakeName", b_name)
                    attr.SetUserString("Surface_Path", str(path))
                    attr.SetUserString("Surface_ID", branch_id)
                    attr.SetUserString("Material_Tag", f_tag)
                    
                    # BIND TO RHINO GROUP
                    attr.AddToGroup(group_idx)
                    
                    doc.Objects.AddMesh(f_mesh, attr)
                
            status_text = "BAKED" if items_replaced == 0 else f"REPLACED ({items_replaced})"
            bake_status = f"\n[ {status_text} TO RHINO ]"
            doc.Views.Redraw()

        # ==========================================================================
        # PASS 4: UI TELEMETRY
        # ==========================================================================
        t_end = time.perf_counter()
        execution_ms = (t_end - t_start) * 1000.0

        ui_lines = [
            "COLORED MESH TILES",
            f"Time: {execution_ms:.2f} ms",
            "---"
        ]
        
        for idx, col in enumerate(gradient_colors):
            count = global_color_counts.get(col, 0)
            ui_lines.append(f"Tile {idx + 1}: {count}")
            
        if has_accent and global_color_counts.get(accent_color, 0) > 0:
            ui_lines.append(f"Accent Tile: {global_color_counts[accent_color]}")
            
        ui_lines.append("---")
        ui_lines.append(f"Total Tiles: {total_panels}{bake_status}")

        ghenv.Component.Message = "\n".join(ui_lines)

File: Components/ColorMeshTiles.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using Grasshopper;
using Grasshopper.Kernel;
using Grasshopper.Kernel.Data;
using Grasshopper.Kernel.Types;
using Rhino;
using Rhino.Geometry;

namespace Enzyme.Components
{
    public class ColorMeshTiles : GH_Component
    {
        public ColorMeshTiles()
          : base("ColorMeshTiles", "ColorMeshTiles",
              "BIM-ready Idempotent Engine. Consolidates Mega-Meshes per surface and binds them into Rhino Groups.",
              "Enzyme", "Facade")
        {
        }

                public override void AddedToDocument(GH_Document document)
        {
            base.AddedToDocument(document);
            if (this.Attributes == null) this.CreateAttributes();

            bool hasSources = false;
            foreach (var param in this.Params.Input)
                if (param.SourceCount > 0) { hasSources = true; break; }

            if (!hasSources)
            {
                var colors = new System.Drawing.Color[] {
                    System.Drawing.Color.FromArgb(240, 120, 120),
                    System.Drawing.Color.FromArgb(200, 120, 120),
                    System.Drawing.Color.FromArgb(250, 210, 210)
                };
                Enzyme.Utils.AutoWireHelper.WireMergeWithSwatches(this, document, 1, colors, 150, -140);
                Enzyme.Utils.AutoWireHelper.WireColorSwatch(this, document, 2, System.Drawing.Color.FromArgb(255, 30, 0), 210, -50);
                Enzyme.Utils.AutoWireHelper.WireValueList(this, document, 3, new string[]{"X", "Y", "Z"}, new string[]{"0", "1", "2"}, 300, -10);
                Enzyme.Utils.AutoWireHelper.WireSlider(this, document, 4, 0.0, 100.0, 30.0, 330, 30);
                Enzyme.Utils.AutoWireHelper.WireSlider(this, document, 5, 0.0, 100.0, 40.0, 330, 70);
                Enzyme.Utils.AutoWireHelper.WireSlider(this, document, 6, 0.0, 1.0, 0.94, 330, 110);
                Enzyme.Utils.AutoWireHelper.WireButton(this, document, 7, 210, 150);
                Enzyme.Utils.AutoWireHelper.WireOutputParam(this, document, 0, "mesh", 220, -120);
                Enzyme.Utils.AutoWireHelper.WireOutputPanel(this, document, 1, 220, -50, 180, 50);
                Enzyme.Utils.AutoWireHelper.WireOutputPanel(this, document, 2, 220, 20, 180, 50);
                Enzyme.Utils.AutoWireHelper.WireOutputParam(this, document, 3, "mesh", 220, 90);
            }
        }

        protected override void RegisterInputParams(GH_InputParamManager pManager)
        {
            pManager.AddCurveParameter("polylines", "P", "Tree of polylines", GH_ParamAccess.tree);
            pManager.AddColourParameter("gradient_colors", "C", "List of gradient colors", GH_ParamAccess.list);
            pManager.AddColourParameter("accent_color", "AC", "Accent color", GH_ParamAccess.item, Color.Empty);
            pManager.AddIntegerParameter("axis", "A", "Axis (0=X, 1=Y, 2=Z)", GH_ParamAccess.item, 2);
            pManager.AddNumberParameter("jitter_pct", "J", "Jitter percentage", GH_ParamAccess.item, 0.0);
            pManager.AddNumberParameter("accent_pct", "AP", "Accent percentage", GH_ParamAccess.item, 0.0);
            pManager.AddNumberParameter("inset_factor", "I", "Inset factor", GH_ParamAccess.item, 1.0);
            pManager.AddBooleanParameter("bake_trigger", "B", "Bake trigger", GH_ParamAccess.item, false);
            pManager.AddTextParameter("bake_name", "BN", "Bake name", GH_ParamAccess.item, "");
            
            pManager[2].Optional = true;
            pManager[3].Optional = true;
            pManager[4].Optional = true;
            pManager[5].Optional = true;
            pManager[6].Optional = true;
            pManager[7].Optional = true;
            pManager[8].Optional = true;
        }

        protected override void RegisterOutputParams(GH_OutputParamManager pManager)
        {
            pManager.AddMeshParameter("panel_mesh", "M", "Consolidated panels", GH_ParamAccess.tree);
            pManager.AddColourParameter("panel_colors", "C", "Panel colors", GH_ParamAccess.tree);
            pManager.AddTextParameter("panel_tags", "T", "Panel tags", GH_ParamAccess.tree);
            pManager.AddCurveParameter("panel_geometry", "G", "Panel geometry curves", GH_ParamAccess.tree);
        }

        protected override void SolveInstance(IGH_DataAccess DA)
        {
            var t_start = System.Diagnostics.Stopwatch.StartNew();

            GH_Structure<GH_Curve> polylinesTree;
            if (!DA.GetDataTree(0, out polylinesTree)) return;
            
            List<Color> gradient_colors = new List<Color>();
            if (!DA.GetDataList(1, gradient_colors)) return;
            
            if (polylinesTree.DataCount == 0 || gradient_colors.Count == 0)
            {
                Message = "COLORED_TILES\nSTATUS: IDLE\n---\nAWAITING DATA";
                return;
            }

            Color accent_color = Color.Empty;
            DA.GetData(2, ref accent_color);
            
            int axis_idx = 2;
            DA.GetData(3, ref axis_idx);
            
            double jitter_pct = 0.0;
            DA.GetData(4, ref jitter_pct);
            double j_factor = Math.Max(0.0, Math.Min(100.0, jitter_pct)) / 100.0;
            
            double accent_pct = 0.0;
            DA.GetData(5, ref accent_pct);
            double a_factor = Math.Max(0.0, Math.Min(100.0, accent_pct)) / 100.0;
            
            double inset_factor = 1.0;
            DA.GetData(6, ref inset_factor);
            double i_factor = Math.Max(0.0, Math.Min(1.0, inset_factor));
            
            bool do_bake = false;
            DA.GetData(7, ref do_bake);
            
            string b_name = "";
            DA.GetData(8, ref b_name);
            
            Random random = new Random(42);
            int num_available_colors = gradient_colors.Count;
            
            Dictionary<Color, string> color_to_tag = new Dictionary<Color, string>();
            for (int i = 0; i < num_available_colors; i++)
            {
                if (!color_to_tag.ContainsKey(gradient_colors[i]))
                    color_to_tag[gradient_colors[i]] = $"Tile {i + 1}";
            }
            
            Dictionary<Color, int> global_color_counts = new Dictionary<Color, int>();
            foreach (var c in gradient_colors)
            {
                if (!global_color_counts.ContainsKey(c))
                    global_color_counts[c] = 0;
            }
                
            bool has_accent = accent_color != Color.Empty && a_factor > 0;
            if (has_accent)
            {
                global_color_counts[accent_color] = 0;
                color_to_tag[accent_color] = "Accent Tile";
            }

            GH_Structure<GH_Mesh> out_mesh_tree = new GH_Structure<GH_Mesh>();
            GH_Structure<GH_Colour> out_cols_tree = new GH_Structure<GH_Colour>();
            GH_Structure<GH_String> out_tags_tree = new GH_Structure<GH_String>();
            GH_Structure<GH_Curve> out_geo_tree = new GH_Structure<GH_Curve>();

            List<double> global_keys = new List<double>();
            foreach (var branch in polylinesTree.Branches)
            {
                foreach (GH_Curve ghCrv in branch.Cast<GH_Curve>())
                {
                    if (ghCrv == null || ghCrv.Value == null) continue;
                    BoundingBox bbox = ghCrv.Value.GetBoundingBox(true);
                    double val = axis_idx == 0 ? bbox.Center.X : (axis_idx == 1 ? bbox.Center.Y : bbox.Center.Z);
                    global_keys.Add(val);
                }
            }

            if (global_keys.Count == 0)
            {
                Message = "HME\nSTATUS: EMPTY\n---\nNO VALID CURVES";
                return;
            }

            double global_min = global_keys.Min();
            double global_max = global_keys.Max();
            double coordinate_span = global_max - global_min;
            if (Math.Abs(coordinate_span) < 1e-6) coordinate_span = 1.0;

            int total_panels = 0;

            for (int i = 0; i < polylinesTree.Branches.Count; i++)
            {
                GH_Path path = polylinesTree.Paths[i];
                var branch = polylinesTree.Branches[i];
                
                out_mesh_tree.EnsurePath(path);
                out_cols_tree.EnsurePath(path);
                out_tags_tree.EnsurePath(path);
                out_geo_tree.EnsurePath(path);
                
                List<GH_Curve> branch_geometries = new List<GH_Curve>();
                
                Dictionary<Color, Mesh> local_mesh_buckets = new Dictionary<Color, Mesh>();
                foreach (var c in gradient_colors)
                {
                    if (!local_mesh_buckets.ContainsKey(c))
                        local_mesh_buckets[c] = new Mesh();
                }
                    
                if (has_accent && !local_mesh_buckets.ContainsKey(accent_color))
                {
                    local_mesh_buckets[accent_color] = new Mesh();
                }

                foreach (GH_Curve ghCrv in branch.Cast<GH_Curve>())
                {
                    if (ghCrv == null || ghCrv.Value == null) continue;
                    
                    Polyline polyline;
                    if (!ghCrv.Value.TryGetPolyline(out polyline)) continue;
                    
                    BoundingBox bbox = ghCrv.Value.GetBoundingBox(true);
                    Point3d center_pt = bbox.Center;
                    
                    if (i_factor < 1.0)
                    {
                        Transform scale_transform = Transform.Scale(center_pt, i_factor);
                        polyline.Transform(scale_transform);
                    }
                    
                    List<Point3d> pts = polyline.ToList();
                    if (pts.Count < 4) continue;
                    if (pts[0].EpsilonEquals(pts[pts.Count - 1], 1e-6))
                    {
                        pts.RemoveAt(pts.Count - 1);
                    }
                    
                    int num_vertices = pts.Count;
                    
                    double current_val = axis_idx == 0 ? center_pt.X : (axis_idx == 1 ? center_pt.Y : center_pt.Z);
                    double t_base = (current_val - global_min) / coordinate_span;
                    
                    if (j_factor > 0)
                    {
                        t_base += (random.NextDouble() * j_factor) - (j_factor * 0.5);
                    }
                    
                    t_base = Math.Max(0.0, Math.Min(0.999999, t_base));
                    int color_index = (int)(t_base * num_available_colors);
                    Color cell_color = gradient_colors[color_index];
                    
                    if (has_accent && random.NextDouble() < a_factor)
                    {
                        cell_color = accent_color;
                    }
                    
                    global_color_counts[cell_color]++;
                    
                    Mesh target_mesh = local_mesh_buckets[cell_color];
                    int v_start_idx = target_mesh.Vertices.Count;
                    
                    target_mesh.Vertices.Add(center_pt);
                    target_mesh.VertexColors.Add(cell_color);
                    
                    foreach (var p in pts)
                    {
                        target_mesh.Vertices.Add(p);
                        target_mesh.VertexColors.Add(cell_color);
                    }
                    
                    for (int j = 0; j < num_vertices; j++)
                    {
                        int next_j = (j + 1) % num_vertices;
                        target_mesh.Faces.AddFace(v_start_idx, v_start_idx + 1 + j, v_start_idx + 1 + next_j);
                    }
                    
                    branch_geometries.Add(new GH_Curve(polyline.ToPolylineCurve()));
                    total_panels++;
                }
                
                List<GH_Mesh> local_meshes = new List<GH_Mesh>();
                List<GH_Colour> local_colors = new List<GH_Colour>();
                List<GH_String> local_tags = new List<GH_String>();
                
                foreach (var kvp in local_mesh_buckets)
                {
                    Color color = kvp.Key;
                    Mesh m_bucket = kvp.Value;
                    
                    if (m_bucket.Vertices.Count > 0)
                    {
                        m_bucket.Normals.ComputeNormals();
                        m_bucket.Compact();
                        local_meshes.Add(new GH_Mesh(m_bucket));
                        local_colors.Add(new GH_Colour(color));
                        local_tags.Add(new GH_String(color_to_tag[color]));
                    }
                }
                
                if (local_meshes.Count > 0)
                {
                    out_mesh_tree.AppendRange(local_meshes, path);
                    out_cols_tree.AppendRange(local_colors, path);
                    out_tags_tree.AppendRange(local_tags, path);
                }
                if (branch_geometries.Count > 0)
                {
                    out_geo_tree.AppendRange(branch_geometries, path);
                }
            }

            DA.SetDataTree(0, out_mesh_tree);
            DA.SetDataTree(1, out_cols_tree);
            DA.SetDataTree(2, out_tags_tree);
            DA.SetDataTree(3, out_geo_tree);

            string bake_status = "";
            int items_replaced = 0;

            if (do_bake)
            {
                var doc = RhinoDoc.ActiveDoc;
                if (doc != null)
                {
                    if (!string.IsNullOrEmpty(b_name))
                    {
                        var existing_objs = doc.Objects.FindByUserString("ElefrontBakeName", b_name, false);
                        if (existing_objs != null && existing_objs.Length > 0)
                        {
                            foreach (var obj in existing_objs)
                            {
                                doc.Objects.Delete(obj.Id, true);
                                items_replaced++;
                            }
                        }
                    }

                    for (int i = 0; i < out_mesh_tree.Branches.Count; i++)
                    {
                        GH_Path path = out_mesh_tree.Paths[i];
                        var branch_meshes = out_mesh_tree.Branches[i];
                        var branch_colors = out_cols_tree.Branches[i];
                        var branch_tags = out_tags_tree.Branches[i];

                        string branch_id = path.ToString().Replace("{", "").Replace("}", "").Replace(";", "_");
                        string group_name = !string.IsNullOrEmpty(b_name) ? $"SurfaceGroup_{b_name}_{branch_id}" : $"SurfaceGroup_{branch_id}";
                        int group_idx = -1;

                        foreach (var g in doc.Groups)
                        {
                            if (g != null && g.Name == group_name)
                            {
                                group_idx = g.Index;
                                break;
                            }
                        }

                        if (group_idx < 0)
                        {
                            group_idx = doc.Groups.Add(group_name);
                        }

                        for (int j = 0; j < branch_meshes.Count; j++)
                        {
                            Mesh f_mesh = branch_meshes[j].Value;
                            Color f_color = branch_colors[j].Value;
                            string f_tag = branch_tags[j].Value;

                            string layer_name = $"HexFacade_{f_tag.Replace(" ", "")}";
                            int layer_idx = doc.Layers.Find(layer_name, true);

                            if (layer_idx < 0)
                            {
                                var new_layer = new Rhino.DocObjects.Layer();
                                new_layer.Name = layer_name;
                                new_layer.Color = f_color;

                                var new_mat = new Rhino.DocObjects.Material();
                                new_mat.DiffuseColor = f_color;
                                new_mat.Name = $"Mat_{layer_name}";

                                int mat_idx = doc.Materials.Add(new_mat);
                                new_layer.RenderMaterialIndex = mat_idx;
                                layer_idx = doc.Layers.Add(new_layer);
                            }

                            var attr = new Rhino.DocObjects.ObjectAttributes();
                            attr.LayerIndex = layer_idx;
                            attr.ColorSource = Rhino.DocObjects.ObjectColorSource.ColorFromLayer;
                            attr.MaterialSource = Rhino.DocObjects.ObjectMaterialSource.MaterialFromLayer;

                            if (!string.IsNullOrEmpty(b_name))
                            {
                                attr.SetUserString("ElefrontBakeName", b_name);
                            }
                            attr.SetUserString("Surface_Path", path.ToString());
                            attr.SetUserString("Surface_ID", branch_id);
                            attr.SetUserString("Material_Tag", f_tag);

                            attr.AddToGroup(group_idx);

                            doc.Objects.AddMesh(f_mesh, attr);
                        }
                    }

                    string status_text = items_replaced == 0 ? "BAKED" : $"REPLACED ({items_replaced})";
                    bake_status = $"\n[ {status_text} TO RHINO ]";
                    doc.Views.Redraw();
                }
            }

            t_start.Stop();
            double execution_ms = t_start.Elapsed.TotalMilliseconds;

            List<string> ui_lines = new List<string>
            {
                "COLORED MESH TILES",
                $"Time: {execution_ms:F2} ms",
                "---"
            };

            for (int i = 0; i < gradient_colors.Count; i++)
            {
                Color col = gradient_colors[i];
                int count = global_color_counts.ContainsKey(col) ? global_color_counts[col] : 0;
                ui_lines.Add($"Tile {i + 1}: {count}");
            }

            if (has_accent && global_color_counts.ContainsKey(accent_color) && global_color_counts[accent_color] > 0)
            {
                ui_lines.Add($"Accent Tile: {global_color_counts[accent_color]}");
            }

            ui_lines.Add("---");
            ui_lines.Add($"Total Tiles: {total_panels}{bake_status}");

            Message = string.Join("\n", ui_lines);
        }

        protected override Bitmap Icon
        {
            get
            {
                return IconLoader.Load("ColorMeshTiles.png");
            }
        }

        public override Guid ComponentGuid
        {
            get { return new Guid("1f3f6c8d-3b56-4c22-b2a8-1234abcd5678"); }
        }
    }
}
